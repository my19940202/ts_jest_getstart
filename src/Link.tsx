import * as React from "react";
import { render } from "react-dom";

import "./styles.css";
const STATUS = {
    HOVERED: 'hovered',
    NORMAL: 'normal',
};

// 单独抽离，便于单测的时候引入测试（如果把index.tsx引入测试会报错。渲染的节点node层渲染找不到）
export default class Link extends React.Component<any, any> {
    constructor(props: any) {
        super(props);
        this._onMouseEnter = this._onMouseEnter.bind(this);
        this._onMouseLeave = this._onMouseLeave.bind(this);
        this._click = this._click.bind(this);
        this._change = this._change.bind(this);

        this.state = {
            class: STATUS.NORMAL,
            counter: 0,
            name: ''
        };
    }

    _onMouseEnter() {
        this.setState({ class: STATUS.HOVERED });
    }

    _onMouseLeave() {
        this.setState({ class: STATUS.NORMAL });
    }

    _click() {
        let tmp = this.state.counter;
        this.setState({ counter: tmp + 1 });
    }

    _change(e: any) {
        console.log(e.target.value, 'e.target.value');
        this.setState({name: e.target.value});
    }

    render() {
        return (
            <>
                <a
                    className={this.state.class}
                    href={this.props.page || '#'}
                    onMouseEnter={this._onMouseEnter}
                    onMouseLeave={this._onMouseLeave}
                >
                    {this.props.children}
                </a>
                <div>
                    <button id="btn" onClick={this._click}>click me</button>
                    counter: <span className="txt">{this.state.counter}</span>
                </div>
                <div>
                    <input onChange={this._change} type="text" value={this.state.name} />
                    name: <span className="name">{this.state.name}</span>
                </div>
            </>
        );
    }
}
