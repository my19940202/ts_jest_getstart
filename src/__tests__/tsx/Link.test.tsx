// Link.react.test.js
import * as React from 'react';
import Link from '../../Link';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

// TODO详细看看每个流程的作用
// 单测运行前的一系列准备工作：https://jestjs.io/docs/zh-Hans/api#afterallfn-timeout
afterAll(() => {
    console.log('afterAll');
});

afterEach(() => {
    console.log('afterEach');
});

beforeAll(() => {
    console.log('beforeAll');
});

beforeEach(() => {
    console.log('beforeEach');
});

// creates a block that groups together several related tests.
describe("Link's tests", () => {
    // 最简单的组件测试，确认不会抛出异常
    test('冒烟测试', () => {
        // shallow render mount 区别搞清楚(https://airbnb.io/enzyme/docs/api/)
        const wrapper = shallow(
            <Link page="http://www.facebook.com">Facebook</Link>,
        );
        expect(wrapper.contains("Facebook")).toEqual(true);
        // 生成快照
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    // it是test的别名(https://jestjs.io/docs/zh-Hans/api#testname-fn-timeout)
    let timeout = 2000;
    it('trigger event in Link Component', () => {
        const wrapper = shallow(
            <Link page="http://www.facebook.com">Facebook</Link>,
        );

        // 点击button counter累加
        wrapper.find('button').simulate('click');
        wrapper.find('button').simulate('click');
        wrapper.find('button').simulate('click');
        expect(wrapper.find('.txt').text()).toEqual('3');

        // TODO: 常用事件
        // change
        wrapper.find('input').simulate('change', {
            target: {value: 'world'}
        });
        expect(wrapper.find('.name').text()).toEqual('world');
        // hover
        wrapper.find('a').simulate('mouseenter');
        expect(wrapper.find('a').hasClass('hovered')).toEqual(true);
        wrapper.find('a').simulate('mouseleave');
        expect(wrapper.find('a').hasClass('normal')).toEqual(true);
    }, timeout); // timeout可选参数
})