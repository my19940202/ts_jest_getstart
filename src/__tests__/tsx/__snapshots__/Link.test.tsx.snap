// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link's tests 冒烟测试 1`] = `
<Fragment>
  <a
    className="normal"
    href="http://www.facebook.com"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Facebook
  </a>
  <div>
    <button
      id="btn"
      onClick={[Function]}
    >
      click me
    </button>
    counter: 
    <span
      className="txt"
    >
      0
    </span>
  </div>
  <div>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    name: 
    <span
      className="name"
    />
  </div>
</Fragment>
`;
